matrix_client/room.py

```diff
@@ -468,14 +500,18 @@ def get_joined_members(self):
         Returns:
             {user_id: {"displayname": str or None}}: Dictionary of joined members.
         """
+        if self._members:
+            return self._members
         response = self.client.api.get_room_members(self.room_id)
-        rtn = {
-            event["state_key"]: {
-                "displayname": event["content"].get("displayname"),
-            } for event in response["chunk"] if event["content"]["membership"] == "join"
-        }
-
-        return rtn
+        for event in response["chunk"]:
+            if event["content"]["membership"] == "join":
+                self._mkmembers({
+                    "user_id": event["state_key"],
+                    "displayname": event["content"].get("displayname")})
+        return self._members
+
+    def _mkmembers(self, member):
+        self._members.append(member)
```

Comment:
Why store `self._members` as a list of dictionaries with just two keys apiece
rather than as a single dictionary? Something like:

```python
{
    "@foo:example.com": "Example Displayname",
    "@bar:example.org": "Another displayname"
}
```

Reply:
or maybe better yet:

```python
{
    "@foo:example.com": <User object>,
    "@bar:example.com": <User object>
}
```

Then all this stuff about keeping track of users should be more natural in
the future. Any caching of displayname, etc. would have a natural place.

Reply:
The `User` depends on `api` and the `displayname` would be than retrieved via `api.get_display_name` this would result in another request. I changed the code for `User` to also set `displayname` as we already retrieved it.